src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbol.cs

```diff
@@ -1263,24 +1243,61 @@ internal override void ForceComplete(SourceLocation locationOpt, CancellationTok
                                     parameter.Type.CheckAllConstraints(conversions, parameter.Locations[0], diagnostics);
                                 }
 
-                                if (_state.NotePartComplete(CompletionPart.Parameters))
+                                if (_state.NotePartComplete(CompletionPart.StartPropertyParameters))
                                 {
                                     this.AddDeclarationDiagnostics(diagnostics);
                                     DeclaringCompilation.SymbolDeclaredEvent(this);
+                                    var completedOnThisThread = _state.NotePartComplete(CompletionPart.FinishPropertyParameters);
+                                    Debug.Assert(completedOnThisThread);
                                 }
 
                                 diagnostics.Free();
                             }
                             else
                             {
-                                if (_state.NotePartComplete(CompletionPart.Parameters))
+                                if (_state.NotePartComplete(CompletionPart.StartPropertyParameters))
                                 {
                                     DeclaringCompilation.SymbolDeclaredEvent(this);
+                                    var completedOnThisThread = _state.NotePartComplete(CompletionPart.FinishPropertyParameters);
+                                    Debug.Assert(completedOnThisThread);
                                 }
                             }
                         }
                         break;
 
+                    case CompletionPart.FinishPropertyParameters:
+                        // since StartPropertyParameters was not the next incomplete part, it must have
+                        // been completed by another thread. Wait for it to finish the parameters.
+                        _state.SpinWaitComplete(CompletionPart.FinishPropertyParameters, cancellationToken);
+                        break;
```

Comment:
Other cases of `case CompletionPart.Finish...` are bundled with the `case CompletionPart.Start...`. Consider doing that here as well, so we don't need the comment explaining this case and so to avoid having two cases that complete this part.